From b87b0681acf0937941c04888b1b77182e0c0e13d Mon Sep 17 00:00:00 2001
From: Douglas Bagnall <douglas.bagnall@catalyst.net.nz>
Date: Wed, 17 Aug 2016 09:30:58 +1200
Subject: [PATCH] perf-tests <= samba-4-2: s/ad_dc_ntvfs/dc/

Signed-off-by: Douglas Bagnall <douglas.bagnall@catalyst.net.nz>
---
 selftest/perf_tests.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/selftest/perf_tests.py b/selftest/perf_tests.py
index d49bdf4..b50ca06 100644
--- a/selftest/perf_tests.py
+++ b/selftest/perf_tests.py
@@ -18,7 +18,7 @@ samba4srcdir = source4dir()
 samba4bindir = bindir()
 
 plantestsuite_loadlist("samba4.ldap.ad_dc_performance.python(ad_dc_ntvfs)",
-                       "ad_dc_ntvfs",
+                       "dc",
                        [python, os.path.join(samba4srcdir,
                                              "dsdb/tests/python/ad_dc_performance.py"),
                         '$SERVER', '-U"$USERNAME%$PASSWORD"',
-- 
2.7.4

